### PR TITLE
via and hasRelation

### PR DESCRIPTION
**Pull request by bscheshirwork**
Each relation must be set once. Not mater duplicate declaration in another model. 
Each model can store only declaration of the his relations.

| Q                    | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes/no
| Fixed issues  | #14246

### CODE REVIEW

File: framework/db/BaseActiveRecord.php [hunk 1]
```diff
@@ -409,6 +410,43 @@ public function hasMany($class, $link)
...
+     *     return $this->hasOne(Item::className(), ['id' => 'item_id']);
+     * }
+     * ```
+     *
+     * @param ActiveRecord|array|callable|string $target the model class who have relation
+     * @param string $relationName relation name in target (case sensitive, without 'get')
+     * @return ActiveQuery
+     */
+    public function hasRelation($target, $relationName){
+        $target instanceof ActiveRecord || $target = Yii::createObject($target);
```
Comment (dynasource) [inline framework/db/BaseActiveRecord.php]: this looks wrong. I'd expect a hasRelation to return a boolean

File: tests/framework/data/ActiveDataProviderTest.php [hunk 2]
```diff
@@ -101,6 +101,21 @@ public function testActiveRelationVia()
         $this->assertCount(2, $items);
     }
 
+    public function testActiveRelationViaHasRelation()
```
Comment (dynasource) [inline tests/framework/data/ActiveDataProviderTest.php]: a new method is introduced in framework/db/BaseActiveRecord.php but a unit test is supplied in ActiveDataProviderTest.php. I'd expect specific unit tests for BaseActiveRecord?

**Review by dynasource — CHANGES_REQUESTED** [review]

File: framework/db/BaseActiveRecord.php (hunk 1 above)
Comment (bscheshirwork) [inline framework/db/BaseActiveRecord.php]: like a hasOne, hasMany?

So... How about `fromRelation` or `viaRelation` (and directly add via into method)?

Comment (bscheshirwork) [inline framework/db/BaseActiveRecord.php]:
```php
class Order extends ActiveRecord
{
    public function getItems(){
        return $this->viaRelation(OrderItem::className(), 'item', 'orderItems');
    }
    public function getOrderItems()
    {
        return $this->hasMany(OrderItem::className(), ['order_id' => 'id']);
    }
}
```

Comment (dynasource) [inline framework/db/BaseActiveRecord.php]: ahh, you took that as a reference, understood. Well, IMHO hasOne & hasMany are not chosen correctly. It should describe what it does, so something like:
- `getRelationQuery()` or `getLinkQuery`

Comment (bscheshirwork) [inline framework/db/BaseActiveRecord.php]: So... This more like a `viaTable` (Table name -> relation name)

Comment (bscheshirwork) [inline framework/db/BaseActiveRecord.php]: i.e. if add to end of method `->via('relationAttributeName')`

Comment (bscheshirwork) [inline framework/db/BaseActiveRecord.php]:
```php
    public function viaRelation($target, $relationName, $via){
        $target instanceof ActiveRecord || $target = Yii::createObject($target);
        /** @var ActiveQuery $relationQuery */
        $relationQuery = $target->getRelation($relationName);
        $relationQuery->primaryModel = $this;
        return $relationQuery->via($via);
    }
```

**Comment by bscheshirwork** [comment]
```php
   public function viaRelation($relationName, $via, callable $callable = null)
   {
       $relation = $this->getRelation($via);
       $target = Yii::createObject($relation->modelClass);
       /** @var ActiveQuery $relationQuery */
       $relationQuery = $target->getRelation($relationName);
       $relationQuery->primaryModel = $this;
       return $relationQuery->via($via, $callable);
   }
```

File: tests/framework/data/ActiveDataProviderTest.php (hunk 2 above)
Comment (bscheshirwork) [reply tests/framework/data/ActiveDataProviderTest.php]: tests/framework/db/ActiveRecordTest.php
```php
    public function testJoinWithViaRelation()
    {
        Order::getDb()->getQueryBuilder()->separator = "\n";
        $rows = Order::find()->joinWith([
            'items2' => function ($q) {
                $q->orderBy('item.id');
            },
        ])->all();
        $this->assertNotEmpty($rows);
    }
```

**Comment by bscheshirwork** [comment]
@dynasource 
>  requested changes on 29 Jul 
all is OK now?

**Comment by dmirogin** [comment]
Hmm, based on your example, what will return getItemsCategory if there are different categories for those items?

File: framework/db/BaseActiveRecord.php [hunk 3]
```diff
@@ -410,6 +411,58 @@ public function hasMany($class, $link)
...
+     * Its signature should be `function($query)`, where `$query` is the query to be customized.
+     * @return ActiveQuery ActiveQuery type of end-side model
+     * @since 2.0.14
+     * @see hasOne()
+     * @see hasMany()
+     */
+    public function viaRelation($relationName, $via, callable $callable = null)
+    {
+        $relation = $this->getRelation($via);
+        $target = Yii::createObject($relation->modelClass);
```
Comment (bscheshirwork) [inline framework/db/BaseActiveRecord.php]: I forgot for new feature `::instance()`  
(PR is older)  
In this place is correct to use it?

**Comment by bscheshirwork** [comment]
@dmirogin thsnks, I will expand a example

**Comment by bscheshirwork** [comment]
Additional question:
What is signature is more user-friendly:

```php
    /**
     * @param string $relationName relation name in target (case sensitive, without 'get')
     * @param string $via relation name in called (self) class (case sensitive, without 'get')
     * @param callable $callable a PHP callback for customizing the relation associated with the junction ActiveQuery in `via`.
     * Its signature should be `function($query)`, where `$query` is the query to be customized.
     * @return ActiveQuery ActiveQuery type of end-side model
     * @since 2.0.14
     * @see hasOne()
     * @see hasMany()
     */
    public function viaRelation($relationName, $via, callable $callable = null)

```
or
 
```php
    /**
     * @param string $via relation name in called (self) class (case sensitive, without 'get')
     * @param string $relationName relation name in target (case sensitive, without 'get')
     * @param callable $callable a PHP callback for customizing the relation associated with the junction ActiveQuery in `via`.
     * Its signature should be `function($query)`, where `$query` is the query to be customized.
     * @return ActiveQuery ActiveQuery type of end-side model
     * @since 2.0.14
     * @see hasOne()
     * @see hasMany()
     */
    public function viaRelation($via, $relationName, callable $callable = null)

```

File: framework/db/BaseActiveRecord.php (hunk 3 above)
Comment (samdark) [reply framework/db/BaseActiveRecord.php]: Yes.

**Comment by samdark** [comment]
How about the following?

```php
public function relationVia($relationName, $via, callable $callable = null)
```

**Comment by bscheshirwork** [comment]
@samdark 
>How about the following?
> public function relationVia($relationName, $via, callable $callable = null)

I try it first. 
I try to write some (~10-15) functions in this notation.
I think this is not user-friendly (have more differences of hasOne/hasMany, not natural)

**Comment by bscheshirwork** [comment]
So... I try to use it with reverse
using `createRelationQuery` is better way.

**Comment by samdark** [comment]
Won't make it into 2.0.

**Pull request closed without merging**